hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -201,15 +208,51 @@ private void handlePlatformTransaction(
             recordBuilder.status(ResponseCodeEnum.UNKNOWN);
         }
 
-        // TODO: handle long scheduled transactions
+        // store all records at once
+        final var recordListResult = recordListBuilder.build();
+        recordManager.endUserTransaction(recordListResult.recordStream());
+
+        // Verify if the transaction is a duplicate and add it to the cache
+        if (preHandleResult != null) { // this should always be true, checking just in case
+            try {
+                checkDuplicatesAndIncludeInCache(
+                        preHandleResult,
+                        recordListResult.mainRecord().recordStreamItem().record(),
+                        consensusNow);
+            } catch (final PreCheckException e) {
+                recordFailedTransaction(e.responseCode(), recordBuilder, recordListBuilder);
+            }
+        }
 
         // TODO: handle system tasks
+    }
 
-        // store all records at once
-        recordManager.endUserTransaction(recordListBuilder.build());
+    private void checkDuplicatesAndIncludeInCache(
+            @NonNull final PreHandleResult preHandleResult,
+            @NonNull final TransactionRecord transactionRecord,
+            @NonNull final Instant consensusNow)
+            throws PreCheckException {
+        final var transactionID = transactionRecord.transactionIDOrThrow();
+        final var foundTransactionRecord = hederaRecordCache.getRecord(transactionID);
+
+        addTransactionToCache(preHandleResult, transactionRecord, consensusNow);
+
+        if (foundTransactionRecord != null) {
+            throw new PreCheckException(DUPLICATE_TRANSACTION);
+        }
+    }
+
+    private void addTransactionToCache(
+            @NonNull final PreHandleResult preHandleResult,
+            @NonNull final TransactionRecord transactionRecord,
+            @NonNull final Instant consensusNow) {
+        final var txBody = preHandleResult.txInfo().txBody();
+        final var nodeId = txBody.nodeAccountID().accountNum();
+        final var payerAccountId = preHandleResult.payer();
+        hederaRecordCache.add(nodeId, payerAccountId, transactionRecord, consensusNow);
```

Comment:
My preference is that for a method like this, that is private, and used in only one place, and is only a few lines of code, is to fold the code back into the calling method and include a comment. I know this opinion differs from engineer to engineer, but for me, personally, this kind of method doesn't have enough importance to be broken out. But that's subjective.

Reply:
I really like these kinds of methods because they are easy to understand and are simplifying the main methods. In this case, if I try to find a bug or something in handlePlatformTransaction I might not be interested in the cache at all and I will just skip addTransactionToCache method.

That being said if this is not how it's done in Hedenra I'm okay with refactoring it. It's better to be consistent. Just let me know and I will delete this method.

Reply:
The method got reduced so much that it really doesn't make much sense to have it as a separate method. I removed the method 👍